lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1229 @@
+
+		//llvm::outs()<<"join\n";
+		//this->print(llvm::outs());
+		//other.print(llvm::outs());
+
+		//Insert the other in this
+		for(const auto& other_range : other.ranges){
+			// We have to insert the other range
+			// If we find it then we have to join, else we have to insert it
+			this->insertOrJoin(other_range.first, other_range.second);
+		}
+		
+		// Do the same for the to_check case
+		for(const auto& other_to_check: other.to_check){
+			this->to_check.insert(other_to_check);
+		}
+
+		//
+		if (*this==copy and other==copy){
+			//this->print(llvm::outs());
+			return mlir::ChangeResult::NoChange;
+		}
+		//this->print(llvm::outs());
+		return mlir::ChangeResult::Change;
+	}
+
+	// Operation to copy two lattices
+	// NB: this is a deep copy
+	mlir::ChangeResult ConstraintsLattice::copy(const ConstraintsLattice& other){
+		
+		//llvm::outs()<<"mine : ";this->print(llvm::outs());
+		//llvm::outs()<<"other: ";other.print(llvm::outs());
+
+		if (*this==other){
+			return mlir::ChangeResult::NoChange;
+		}
+		this->ranges=other.ranges;
+		this->to_check=other.to_check;
+		return mlir::ChangeResult::Change;
+	}
+
+	// Weird stuff but it works 
+	mlir::ChangeResult ConstraintsLattice::copy(const mlir::dataflow::AbstractDenseLattice& val){
+		return this->copy(*static_cast<const ConstraintsLattice*>(&val));
+	}
+
+	// Printing current lattice (only for debugging)
+	void ConstraintsLattice::print(raw_ostream& os) const{
+
+		os<<"Ranges:\n";
+		if(not this->ranges.empty()){
+			for(auto range: this->ranges){
+				os << "SSA: "<< range.first.first <<" Branch: "<< range.first.second << 
+				" RANGE: (" <<range.second.getLower().getSExtValue() << ", " << range.second.getUpper().getSExtValue() << ")\n";
+			}
+		}
+		else {
+			os<<"Empty\n";
+		}
+		os<<"To check:\n";
+		if(not this->to_check.empty()){
+			for(auto check: this->to_check){
+				os << "SSA: "<< check << "\n";
+			}
+		}
+		else {
+			os<<"Empty\n";
+		}
+	}
+
+	bool ConstraintsLattice::operator==(const ConstraintsLattice& other) const{
+		return this->ranges == other.ranges
+			and this->to_check == other.to_check;
+	}
+
+	// Operation called when visiting a region terminator
+	bool ConstraintsLattice::setCurrentBranch(mlir::RegionBranchOpInterface* branchOp){
+		// The yield has this format
+		// %12 = rlc.constant true !rlc.bool
+		// rlc.yield %12 : !rlc.bool
+		// So I need to retrieve "true" in this case
+
+		// So here we need to be very careful 
+		// We could also directly return something that is not a constant
+		// For instant we could do 'return a>10' but in this case we can just call the function that handles it
+		// Here we get the rlc.constant true
+		const auto& value= mlir::cast<mlir::rlc::Yield>(branchOp->getOperation()).getArguments()[0];
+
+		//value.getDefiningOp()->print(llvm::outs());
+
+		// Case when I am directly returning an operation
+		// TODO: wrap in some static method or put a trait
+		if( mlir::isa<mlir::rlc::GreaterOp>(value.getDefiningOp()) or	
```

Comment:
constraint lattice cannot know the exhaustive list of operations to handle here, because they may change.